simulator/infra/cache/cache_tag_array.h

```diff
@@ -16,41 +16,26 @@
 #include <infra/log.h>
 #include <infra/macro.h>
 #include <infra/types.h>
+#include <infra/replacement/cache_replacement.h>
 
 // Replacement algorithm modules (LRU)
-class LRUCacheInfo
-{
-    public:
-        explicit LRUCacheInfo( std::size_t ways);
-
-        void touch( std::size_t way);
-        std::size_t update();
-        std::size_t get_ways() const { return ways; }
-
-    private:
-        std::list<std::size_t> lru_list{};
-        std::unordered_map<std::size_t, decltype(lru_list.cbegin())> lru_hash{};
-
-        const std::size_t ways;
-};
 
-class LRUModule
+class ReplacementModule
 {
     public:
-        LRUModule( std::size_t number_of_sets, std::size_t number_of_ways)
-            : lru_info( number_of_sets, LRUCacheInfo( number_of_ways))
-        { }
+        ReplacementModule( std::size_t number_of_sets, std::size_t number_of_ways, int replacement_policy = LRU);
+        ~ReplacementModule() { replacement_info.clear(); }
 
-        void touch( uint32 num_set, uint32 num_way) { lru_info[ num_set].touch( num_way); }
-        auto update( uint32 num_set) { return lru_info[ num_set].update(); }
+        void touch( uint32 num_set, uint32 num_way) { replacement_info[ num_set]->touch( num_way); }
+        auto update( uint32 num_set) { return replacement_info[ num_set]->update(); }
 
     private:
-        std::vector<LRUCacheInfo> lru_info;
+        std::vector<CacheReplacementInterface*> replacement_info;
```

Comment:
Please use `std::unique_ptr` instead of C-style pointer to simplify memory management.